dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,313 @@
+  std::enable_if_t<!has_channels_, int> NumChannels() {
+    return 1;
+  }
+
+  T* data_ = nullptr;
+  int start_ = 0;
+  int end_ = 0;  ///< next empty element
+  int elements_ = 0;
+  int length_ = 0;
+  int num_channels_ = 0;
+};
+
+template <typename T, bool has_channels>
+void load_pixel_with_border(CyclicPixelWrapper<T, has_channels>& cpw, const T* in_ptr, int in_idx,
+                            int stride, int axis_size, span<const T> fill_value) {
+  cpw.PushPixel(in_ptr + boundary::idx_reflect_101(in_idx, axis_size) * stride);
+}
+
+template <typename T, bool has_channels>
+void load_pixel_no_border(CyclicPixelWrapper<T, has_channels>& cpw, const T* in_ptr, int in_idx,
+                          int stride) {
+  cpw.PushPixel(in_ptr + in_idx * stride);
+}
+
+constexpr bool is_convolution_inner_loop(int dim, int ndim, bool has_channels) {
+  if (has_channels) {
+    return dim == ndim - 1;
+  } else {
+    return dim == ndim;
+  }
+}
+template <int dim, int ndim, bool has_channels>
+using is_convolution_inner = std::enable_if_t<is_convolution_inner_loop(dim, ndim, has_channels)>;
+
+template <int dim, int ndim, bool has_channels>
+using is_convolution_outer = std::enable_if_t<!is_convolution_inner_loop(dim, ndim, has_channels)>;
+
+// we're in channel dim
+template <int axis, bool has_channels, int dim = 0, typename Out, typename In, typename W, int ndim>
+is_convolution_inner<dim, ndim, has_channels> ConvolutionCpuImpl(
+    Out* out, const In* in, const W* window, const TensorShape<ndim>& shape,
+    const TensorShape<ndim>& strides, int d, int64_t offset, span<const In> border_fill,
+    In* input_window_buffer, span<W> pixel_tmp, W scale = 1) {
+  auto pixel_stride = strides[axis];
+  auto axis_size = shape[axis];
+  auto num_channels = has_channels ? shape[ndim - 1] : 1;  // channel-last is assumed
+  int r = (d - 1) / 2;                                     // radius = (diameter - 1) / 2
+  // offset <- start of current axis
+  auto* out_ptr = out + offset;
+  auto* in_ptr = in + offset;
+
+  CyclicPixelWrapper<In, has_channels> input_window(input_window_buffer, d, num_channels);
+
+  int in_idx = -r, out_idx = 0;
+  for (in_idx = -r; in_idx < 0; in_idx++) {
+    load_pixel_with_border(input_window, in_ptr, in_idx, pixel_stride, axis_size, border_fill);
+  }
+  if (r < axis_size) {
+    // we load the window without the last element
+    for (; in_idx < r; in_idx++) {
+      load_pixel_no_border(input_window, in_ptr, in_idx, pixel_stride);
+    }
+    for (; out_idx < axis_size - r; out_idx++, in_idx++) {
+      // we load last element of the input window corresponding to the out_idx
+      load_pixel_no_border(input_window, in_ptr, in_idx, pixel_stride);
+      // we have both windows as almost-contiguous buffers
+      input_window.CalculateDot(pixel_tmp.data(), window);
+      for (int c = 0; c < num_channels; c++) {
+        out_ptr[out_idx * pixel_stride + c] = ConvertSat<Out>(pixel_tmp[c] * scale);
+      }
```

Comment:
Maybe you can extract this to a separate function or add scaling to `CalculateDot`?